### PR TITLE
[asset backfill] Fix 'all partitions' targeting

### PR DESCRIPTION
**Pull request by clairelin135**
Similar to https://github.com/dagster-io/dagster/issues/16952.

In certain situations when the `allPartitions = True` flag is applied to an asset backfill, some downstream partitions hang and never get executed.

For example, in the following situation:
- Asset A, Daily starting 2023
- Asset B, Daily starting 2022 (dependent on A) 

The following happens:
1. Backfill is launched with `allPartitions = True`, causing all partitions of A and B are targeted
2. Backfill requests roots -- before the changes in this PR, that would be all partitions of A
3. In subsequent iterations, partitions downstream of A partitions get kicked off once A finishes executing. This would be partitions in B after 2023
4. 2022 partitions in B hang indefinitely

Instead, we should target 2022 partitions of B as part of the 'request roots' call.

This PR fixes this issue by updating the `get_target_root_asset_partitions` call to also find partitions that do not have parents in the target subset. Previously, it only found assets that do not have parents in the target subset.

### CODE REVIEW

**Comment by clairelin135** [comment]
Current dependencies on/for this PR:
* master
  * **PR #17278** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/17278" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/17278?utm_source=stack-comment).

File: python_modules/dagster/dagster/_core/definitions/asset_graph_subset.py [hunk 1]
```diff
@@ -42,7 +42,9 @@ def non_partitioned_asset_keys(self) -> AbstractSet[AssetKey]:
 
     @property
     def asset_keys(self) -> AbstractSet[AssetKey]:
-        return self.partitions_subsets_by_asset_key.keys() | self._non_partitioned_asset_keys
+        return {
```
Comment (clairelin135) [inline python_modules/dagster/dagster/_core/definitions/asset_graph_subset.py:45]: Fixes a bug I discovered -- a partitioned asset key should not be considered part of the subset if its partitions subset is empty

File: python_modules/dagster/dagster/_core/execution/asset_backfill.py [hunk 2]
```diff
@@ -186,11 +204,46 @@ def get_target_root_asset_partitions(self) -> Iterable[AssetKeyPartitionKey]:
...
+
+            previous_root_and_downstream_partitions = root_and_downstream_partitions
+            root_and_downstream_partitions = (
+                root_and_downstream_partitions
+                | _get_self_and_downstream_targeted_subset(unreachable_target_root_partitions)
+            )
+
+        if root_and_downstream_partitions == previous_root_and_downstream_partitions:
+            raise DagsterInvariantViolationError(
+                "Unable to determine root partitions for backfill. The following asset partitions"
```
Comment (clairelin135) [inline python_modules/dagster/dagster/_core/execution/asset_backfill.py:240]: Raises an error when certain partitions will never be targeted, so that we can detect this early on

**Comment by sryza** [comment]
@OwenKephart will your AMP memory changes interact with this?

File: python_modules/dagster/dagster/_core/execution/asset_backfill.py [hunk 3]
```diff
@@ -175,7 +175,25 @@ def have_all_requested_runs_finished(self) -> bool:
 
         return True
 
-    def get_target_root_asset_partitions(self) -> Iterable[AssetKeyPartitionKey]:
+    def get_target_root_asset_partitions(
+        self, instance_queryer: CachingInstanceQueryer
+    ) -> Iterable[AssetKeyPartitionKey]:
+        def _get_self_and_downstream_targeted_subset(
```
Comment (OwenKephart) [inline python_modules/dagster/dagster/_core/execution/asset_backfill.py:181]: is there a way to get a before/after measurement of the performance of this on a larger graph w/ lots of partitions? This example: https://github.com/dagster-io/dagster/blob/master/python_modules/dagster-test/dagster_test/toys/auto_materializing/large_graph.py might be a good potential one to try it out on.

My concern here is that `bfs_filter_asset_partitions` requires you to individually partition map every single partition key in the subset, which can take a very long time compared to just mapping from a large set of keys to the large set of downstream keys.

I was thinking we could instead just use `bfs_filter_subsets` with a condition function of "this asset key is one of the asset keys targeted by the backfill". This would potentially end up with a subset that contains more partitions than we want (as it's not guaranteed that all children partitions will be part of the target subset, even if some partitions for that key are), but we could filter those extra partitions out after the fact by just taking the intersection between the target subset and the subset we get from the process described above. That might require adding another operation to the asset graph subset similar to: https://sourcegraph.com/github.com/dagster-io/dagster/-/blob/python_modules/dagster/dagster/_core/definitions/asset_graph_subset.py?L142, or whatever makes sense to you.

My expectation is that the current implementation would take quite awhile on the example repo, but the bfs_filter_subsets would be significantly faster (open to being proven wrong on either count there)

File: python_modules/dagster/dagster/_core/execution/asset_backfill.py [hunk 4]
```diff
@@ -186,11 +204,46 @@ def get_target_root_asset_partitions(self) -> Iterable[AssetKeyPartitionKey]:
...
-        return list(
+        root_partitions = set(
             self.target_subset.filter_asset_keys(
                 assets_with_no_parents_in_target_subset
             ).iterate_asset_partitions()
         )
+        root_and_downstream_partitions = _get_self_and_downstream_targeted_subset(root_partitions)
+        previous_root_and_downstream_partitions = None
+
+        while (
```
Comment (OwenKephart) [inline python_modules/dagster/dagster/_core/execution/asset_backfill.py:215]: Took me awhile to understand why we needed a loop at all, so summarizing my understanding here and let me know if it's accurate.

My first instinct was that we have:

`root_partitions`:  The set of asset partitions which are part of assets which have no targeted parents.

`root_and_downstream_partitions`: `downstream(root_partitions) & target_subset`, where `downstream(x)` represents the set of asset partitions which are in x or downstream (directly or transitively) of an asset partition in x

I originally that the set we want is `root_partitions | { any target asset partition which is not in downstream(root_partitions) }`, as this will represent any asset partition which has no parents which will be materialized by the backfill. We could then get `{ any target asset partition which is not in downstream(root_partitions) }` just by taking `target_asset_partitions - root_and_downstream_partitions`, so no need for any looping.

However, this fails in the following case:

`A[p1] -> B[p1, p2] -> C[p1, p2]` (i.e. A has a single partition but its downstreams both have two partitions).

`target_asset_partitions` = `<all of them>`
`root_partitions` = `A[p1]`
`root_and_downstream_partitions` = `A[p1], B[p1], C[p1]`
`target_asset_partitions - root_and_downstream_partitions` = `B[p2], C[p2]`.

In this case, that would mean that we would think it'd be ok to launch individual runs for both B and C's p2 partition, whereas in reality we want C's p2 partition to only execute once B's p2 partition is completed.

So instead, we're just adding in the new "roots" of `target_asset_partitions - root_and_downstream_partitions`, then repeating the process to find out what's unreachable from there.

It seems like there _could_ (and that's a big "could") be a more efficient way of doing this (which avoids having to repeatedly call this _get_self_and_downstream_targeted_subset function), but it's a very tricky problem, and you'd need to have a fairly degenerate case in order to need to do a ton of loops here (something like `A[p1] -> B[p1, p2] -> C[p1, p2, p3] -> ...`)

Long story short, I think this loop could use some more comments, and my main concern is still just around perf. The case where we need this special logic is fairly rare, so I want to make sure:

a) that it isn't unusably slow in that case

and

b) if we're not in that case, there's not a significant perf degradation

File: python_modules/dagster/dagster/_core/execution/asset_backfill.py (hunk 2 above)
Comment (OwenKephart) [reply python_modules/dagster/dagster/_core/execution/asset_backfill.py:240]: Is my understanding correct that it's theoretically impossible to hit this case and this is basically just to prevent infinite loops in case something weird happens?

File: python_modules/dagster/dagster/_core/definitions/asset_graph_subset.py (hunk 1 above)
Comment (OwenKephart) [reply python_modules/dagster/dagster/_core/definitions/asset_graph_subset.py:45]: good call

**Review by OwenKephart — CHANGES_REQUESTED** [review]
Overall, this is looking good in structure, but I'm worried about the perf implications. Happy to be proven wrong but have some suggestions in the PR otherwise.

**Comment by OwenKephart** [comment]
> @OwenKephart will your AMP memory changes interact with this?

@sryza hm not seeing any connection between this PR and those changes -- any specific concerns there?

**Comment by sryza** [comment]
> @sryza hm not seeing any connection between this PR and those changes -- any specific concerns there?

Oops re-read this and realized it's not related. I was pattern-matching on the `allPartitions` and the new `all_partitions` argument on the new backfill skip rule and the "hang" situation Claire covered in the description.

File: python_modules/dagster/dagster/_core/execution/asset_backfill.py (hunk 3 above)
Comment (clairelin135) [reply python_modules/dagster/dagster/_core/execution/asset_backfill.py:181]: Thanks for catching this, your concerns about perf are very justified given that the original implementation took over 5 minutes to run the function on the large asset graph 😅 

I switched to `bfs_filter_subsets` -- thanks for the pointer, didn't know this function existed. On the large asset graph it takes less than 15 seconds, which I think is ok given that this is a one-time cost per-backfill and backfills are typically long running anyway

File: python_modules/dagster/dagster/_core/execution/asset_backfill.py (hunk 4 above)
Comment (clairelin135) [reply python_modules/dagster/dagster/_core/execution/asset_backfill.py:217]: > A[p1] -> B[p1, p2] -> C[p1, p2]

Yes, that's pretty much the situation I'm thinking of above.

Agree there could be a more efficient way of having this loop, which I'm interested in improving in the future but in the interest of time I'd like to get this change out ASAP.

File: python_modules/dagster/dagster/_core/execution/asset_backfill.py (hunk 2 above)
Comment (clairelin135) [reply python_modules/dagster/dagster/_core/execution/asset_backfill.py:246]: yeah that's right, assuming the logic in this function is correct the user should never hit this.

I added it because it would be quite awful if the daemon hung forever on this infinite loop

**Review by OwenKephart — APPROVED** [review]
very nice
